hs/src/Reach/Compiler.hs

```diff
@@ -71,337 +72,359 @@ compile env (CompilerOpts {..}) = do
+                -- there are forks, races, timeouts, loops, and all that exciting
+                -- stuff.
+                --
+                -- Producing C is the harder part. There we need to discover what the
+                -- states are and create each one of the state transitions.
+                --
+                pil <- epp flap
+                -- The C program is a state machine... we can display it as dot
+                showp "state.dot" $ stateDiagram pil
+                showp "pil" pil
+                -- In the DL code generation, APIs are represented as participant
+                -- classes, but we don't want them to have to be attached for the
+                -- entire program. So this pass changes an "A"-like program from
+                --
+                -- recv0
+                -- let x = ...
+                -- recv1
+                -- let y = ...
+                -- recv2
+                -- let dom = ...
+                -- send3(Buy(dom))
+                -- ... x, y, dom ...
+                -- recv3
+                -- recv4
+                --
+                -- into
+                --
+                -- let (x, y) = getState
+                -- let dom = ...
+                -- send3(Buy(dom))
+                -- ... x, y, dom ...
+                -- recv3
+                --
+                apc <- apicut pil
+                showp "apc" apc
+                -- We optimize again at this point, because some values may only be of
+                -- interested to some participants, so the A/B/C programs can become
+                -- smaller
+                pl <- bigopt (showp, "pl") apc
+                showp "pl" pl
+                -- Next, we generate the backend code for each one of the connectors
+                -- the user asked for. This return "connector info" structures that
+                -- basically have the bytecode in them, plus stuff the runtime needs.
+                --
+                -- This only looks at the "C" piece
+                let runConnector c = do
+                      let n = conName c
+                      loud $ "running connector " <> show n
+                      conGen c woutnMay $ plp_cpp pl
+                crs <- mapM runConnector dlo_connectors
+                -- Those connector info things will be given to the JS code to get
+                -- included in the actual backend.
+                --
+                -- You might think this only looks at the A & B pieces, but it also has
+                -- to look at the C piece, so we can generate client-only View
+                -- implementations, because Algorand doesn't have a good way to call
+                -- read-only methods of contracts.
+                loud $ "running backend js"
+                backend_js woutn crs $ plp_epp pl
+                return $ crs
+          m_compile_prog :: String -> SLVal -> IO ConnectorObject
+          m_compile_prog which = \case
```

Comment:
I don't like how big this diff is and I feel like I would have done it by making another top level function.

It is weird that the thing called by the cli returns the object rather than that being a sub thing or the object just being set as an effect

It is really weird that this knows about SL values

Reply:
I believe there needs to exist a compile function that takes the SLVal as opposed to the name in the top env, since it is possible to define a Reach.App in any scope. Is there a simpler way I'm not thinking about?

Reply:
Inside of Compiler.hs, we basically are just waiting for a DLProg and then we can go from there. `compileDApp`, which is really part of `Eval` knows how to turn a name into a `DLProg`.

So, I think that `Compiler.hs` should be giving the `Eval` module the `DLProg -> IO ConnectorInfo` function and then Eval will deal with that. I basically imagine that `prepareDAppCompiles` will receive that function too and then the `compileDApp : String -> DLProg` will be used in a really trivial way in Compiler.hs